### PR TITLE
Frogger game

### PR DESCRIPTION
**Pull request by ORybka**
### Object-Oriented JavaScript

[Demo](https://orybka.github.io/frogger-game/)
[Code base](https://github.com/ORybka/frogger-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/ORybka/frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  this.sprite = 'images/char-cat-girl.png';
+  this.x = x;
+  this.y = y;
+  this.width = param.playerWidth;
+  this.height = param.playerHeight;
+  this.startFieldX = param.fieldBorderX[0];
+  this.endFieldX = param.fieldBorderX[1];
+  this.startFieldY = param.fieldBorderY[0];
+  this.endFieldY = param.fieldBorderY[1];
+  this.enemies = allEnemies;
```
Comment (OleksiyRudenko) [inline submissions/ORybka/frogger-game/app.js:49]: A class should be agnostic of the outer world directly. Whatever it needs to know about outer world should be explicitly given to it at construction time. Just like `x` and `y` for it to appear at.

`allEnemies` is somethig from the outer world. Fix this.

Although, it is OK if class refers to global constants (that is their purpose - to be global) and engine API (`Resources` in this case).

File: submissions/ORybka/frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+Player.prototype.loseMessage = function () {
+  setTimeout(() => {
+    alert('Oh no! You lose!');
+    this.resetGame();
+  }, 0);
+};
+
+Player.prototype.resetGame = function () {
+  this.x = 200;
+  this.y = 380;
```
Comment (OleksiyRudenko) [inline submissions/ORybka/frogger-game/app.js:118]: `200` and `300` have semantic meaning. I assume these are some initial position. Would be good to **read** it here.

Make sure all mysterious numbers have descriptive names (defined as constants) across the entire codebase. Many of them already do, just need to make this approach consistent.

File: submissions/ORybka/frogger-game/app.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      }
+      break;
+    case 'down':
+      this.y += this.height;
+      if (this.y >= this.endFieldY) {
+        this.y = this.endFieldY;
+      }
+      break;
+    default:
+      break;
```
Comment (OleksiyRudenko) [inline submissions/ORybka/frogger-game/app.js:90]: Why do we need this here? It does really nothing

File: submissions/ORybka/frogger-game/app.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  }
+};
+
+Player.prototype.checkCollision = function () {
+  for (let i = 0; i < this.enemies.length; i++) {
+    if (this.y <= this.enemies[i].y + this.enemySize && this.y >= this.enemies[i].y - this.enemySize && this.x <= this.enemies[i].x + this.enemySize && this.x >= this.enemies[i].x - this.enemySize) {
+      this.loseMessage();
+    }
+  }
```
Comment (OleksiyRudenko) [inline submissions/ORybka/frogger-game/app.js:99]: How about `.forEach`? It would make the code cleaner and easier to read.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@ORybka a really good job!
Let's make it perfect.

**Comment by ORybka** [comment]
Thank you very much for the comments! :)
I am very glad with myself that I didn't have a lot of them, so finally I learnt some OOP :) 
I hope I understood right and took into consideration all the comments.

**Review by OleksiyRudenko — APPROVED** [review]
@ORybka perfectly done!

**Pull request merged**